pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java

```diff
@@ -96,19 +97,66 @@ private static boolean isLeafPlan(DispatchablePlanMetadata metadata) {
   }
 
   private void assignWorkersToLeafFragment(PlanFragment fragment, DispatchablePlanContext context) {
-    DispatchablePlanMetadata metadata = context.getDispatchablePlanMetadataMap().get(fragment.getFragmentId());
-    // table scan stage, need to attach server as well as segment info for each physical table type.
-    List<String> scannedTables = metadata.getScannedTables();
-    String logicalTableName = scannedTables.get(0);
-    Map<String, RoutingTable> routingTableMap = getRoutingTable(logicalTableName, context.getRequestId());
-    if (routingTableMap.size() == 0) {
-      throw new IllegalArgumentException("Unable to find routing entries for table: " + logicalTableName);
+    // NOTE: For pipeline breaker, leaf fragment can also have children
+    for (PlanFragment child : fragment.getChildren()) {
+      assignWorkersToNonRootFragment(child, context);
+    }
+
+    TableScanNode tableScanNode = findTableScanNode(fragment.getFragmentRoot());
+    Preconditions.checkState(tableScanNode != null, "Failed to find table scan node under leaf fragment");
+    String tableName = tableScanNode.getTableName();
+
+    // Extract partitionKey and numPartitions from hint if provided
+    Map<String, String> tableHintOptions =
+        tableScanNode.getNodeHint()._hintOptions.get(PinotHintOptions.TABLE_HINT_OPTIONS);
```

Comment:
nit: can we simplify all these by putting the hint on the dispatchable plan metadata map during visitor pattern?

Reply:
Moved to `DispatchablePlanVisitor`